File: server/datasets/java_source-code/Multiple-F2/organic_19_bst.java
import java.util.ArrayList;
import java.util.List;

class CustomNode_18 {
    int item_val;
    CustomNode_18 left;
    CustomNode_18 right;
    CustomNode_18(int item_val) { this.item_val = item_val; }
}

public class BST {
    CustomNode_18 root;
    // Standard iterative/recursive BST
public void insert(int val) {
        if (root == null) {
            root = new CustomNode_18(val);
            return;
        }
        CustomNode_18 curr = root;
        while (true) {
            if (val < curr.item_val) {
                if (curr.left == null) {
                    curr.left = new CustomNode_18(val);
                    break;
                }
                curr = curr.left;
            } else {
                if (curr.right == null) {
                    curr.right = new CustomNode_18(val);
                    break;
                }
                curr = curr.right;
            }
        }
    }

    public boolean search(int val) {
        CustomNode_18 curr = root;
        while (curr != null) {
            if (curr.item_val == val) return true;
            curr = (val < curr.item_val) ? curr.left : curr.right;
        }
        return false;
    }

    public List<Integer> inorder() {
        List<Integer> res = new ArrayList<>();
        traverse(root, res);
        return res;
    }

    private void traverse(CustomNode_18 node, List<Integer> res) {
        if (node != null) {
            traverse(node.left, res);
            res.add(node.item_val);
            traverse(node.right, res);
        }
    }
}

File: server/datasets/python_source-code/Multiple-F2/organic_19_bst.py
# Organic BST Student Submission 19
def persistent_insert(node, val):
    if node is None:
        return (val, None, None)
    v, left, right = node
    if val < v:
        return (v, persistent_insert(left, val), right)
    elif val > v:
        return (v, left, persistent_insert(right, val))
    return node

def persistent_search(node, val):
    while node:
        v, left, right = node
        if v == val: return True
        node = left if val < v else right
    return False

def persistent_inorder(node):
    if not node: return []
    v, left, right = node
    return persistent_inorder(left) + [v] + persistent_inorder(right)
